lib/datasets.go

```diff
@@ -489,88 +490,85 @@ func (m *DatasetMethods) Save(p *SaveParams, res *reporef.DatasetRef) error {
-	} else if errors.Is(err, dsref.ErrNotHumanFriendly) {
+		dsf.Assign(ds)
+		ds = dsf
+	}
+
+	if p.Ref == "" && ds.Name != "" {
+		p.Ref = fmt.Sprintf("me/%s", ds.Name)
+	}
+
+	resolver, err := m.inst.resolverForMode("local")
+	if err != nil {
 		return err
-	} else if err != nil {
-		// If some other parse error happened, describe a valid dataset name.
-		return dsref.ErrDescribeValidName
 	}
 
-	// Validate that username is our own, it's not valid to try to save a dataset with someone
-	// else's username. Without this check, base will replace the username with our own regardless,
-	// it's better to have an error to display, rather than silently ignore it.
 	pro, err := m.inst.repo.Profile()
 	if err != nil {
 		return err
 	}
-	if ref.Username != "" && ref.Username != "me" && ref.Username != pro.Peername {
-		return fmt.Errorf("cannot save using a different username than \"%s\"", pro.Peername)
-	}
 
-	// Parsed human-friendly dsref can only have username and name.
-	datasetRef := reporef.DatasetRef{
-		Peername: ref.Username,
-		Name:     ref.Name,
+	ref, isNew, err := base.PrepareSaveRef(ctx, pro, m.inst.logbook, resolver, p.Ref, ds.BodyPath, p.NewName)
+	if err != nil {
+		return err
 	}
 
-	ds := &dataset.Dataset{}
+	success := false
+	rollback := func() {
+		// if creating a new dataset fails, or we're doing a dry-run on a new dataset
+		// we need to remove the dataset
+		if isNew && (!success || p.DryRun) {
+			log.Debugf("removing unused log for new dataset %s", ref)
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second*2)
+			if err := m.inst.logbook.RemoveLog(ctx, ref); err != nil {
+				log.Errorf("couldn't cleanup unused reference: %q", err)
+			}
+			cancel()
+		}
+	}
+	defer rollback()
 
-	// TODO(dustmop): In the future, resolve the dataset ref early, get the initID, and use that
-	// everywhere. If dataset has no name, don't bother to canonicalize, instead, call Infer
-	// here. If either ref failed to resolve, or Infer was called, generate a new initID using
-	// logbook immediately. Regardless, stop using the dsref after this point.
+	ds.Name = ref.Name
+	ds.Peername = ref.Username
 
-	// Check if the dataset has an FSIPath, which requires a different save codepath.
-	err = repo.CanonicalizeDatasetRef(m.inst.repo, &datasetRef)
-	// Ignore errors that happen when saving a new dataset for the first time
-	if err == repo.ErrNotFound || err == repo.ErrEmptyRef {
-		// do nothing
-	} else if err == nil || err == repo.ErrNoHistory {
-		// When saving in an FSI directory, the ref should exist (due to `qri init`), and we
-		// need to load the previous version from the working directory.
-		if datasetRef.FSIPath != "" {
-			ds, err = fsi.ReadDir(datasetRef.FSIPath)
+	var fsiPath string
+	if !isNew {
+		// check for FSI linked data
+		fsiRef := ref.Copy()
+		if err := m.inst.fsi.ResolvedPath(&fsiRef); err == nil {
+			fsiPath = strings.TrimPrefix(fsiRef.Path, "/fsi")
```

Comment:
This `TrimPrefix` call appears in 3 places in this file, could we get a short helper function `refPathToFilesysPath` to collect the logic in one place.

Reply:
sounds good. I've added `fsi.FilesystemPathToLocal` for this purpose